### PR TITLE
Refactor deploy to mavencentral

### PR DESCRIPTION
**Pull request by willynganga**

### CODE REVIEW

File: .github/workflows/release-snapshot.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+      - name: Gradle build
+        uses: gradle/gradle-build-action@842c587ad8aa4c68eeba24c396e15af4c2e9f30a
+        with:
+          arguments: build
+
+      - name: Deploy artifacts
+        id: deploy
+        if: github.ref == 'refs/heads/main'
+        run: ./gradlew publishToSonatype --stacktrace
```
Comment (Thilo-Acrolinx) [inline .github/workflows/release-snapshot.yml:35]: can we use the gradle-build-action here too for consistency?

File: .github/workflows/release.yml
Comment (Thilo-Acrolinx) [inline .github/workflows/release.yml:1]: maybe rename this file to "release-stable.yml" for consistency?

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -70,13 +42,14 @@ jobs:
       - name: Deploy artifacts
         id: deploy
         if: github.ref == 'refs/heads/main'
-        run: ./deploy.sh
+        run: |
+          ./gradlew publishToSonatype closeAndReleaseStagingRepository --stacktrace
```
Comment (Thilo-Acrolinx) [inline .github/workflows/release.yml:46]: Can we use the gradle-build-action here too?

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -70,13 +42,14 @@ jobs:
...
 
       - name: Bump version and push tag
         uses: anothrNick/github-tag-action@1.67.0
         if: steps.deploy.outputs.RELEASE == 'true' && github.ref == 'refs/heads/main'
```
Comment (Thilo-Acrolinx) [inline .github/workflows/release.yml:50]: if we delete the deploy.sh file, who is going to set this RELEASE=true now?

Same for the "Deploy Javadoc" step below

File: build.gradle [hunk 4]
```diff
@@ -8,16 +8,15 @@ plugins {
 	id 'maven-publish'
 	id 'signing'
 	id 'org.ajoberstar.grgit' version '4.1.1'
-	id 'io.codearte.nexus-staging' version '0.30.0'
-	id 'de.marcphilipp.nexus-publish' version '0.4.0'
-	id 'com.github.node-gradle.node' version '7.0.1'
+	id "io.github.gradle-nexus.publish-plugin" version '1.3.0'
+	id 'com.github.node-gradle.node' version '7.0.0'
```
Comment (Thilo-Acrolinx) [inline build.gradle:12]: why downgrade this plugin version from 7.0.1 to 7.0.0?

File: build.gradle [hunk 5]
```diff
@@ -137,7 +122,7 @@ task copyServerSelectorResources(type: Copy, dependsOn: npm_install) {
 	into('src/main/resources/server-selector')
 }
 
-task convertLocalizationEncoding() {
+tasks.register('convertLocalizationEncoding') {
```
Comment (Thilo-Acrolinx) [inline build.gradle:125]: since this change seems unrelated to the switch to another gradle plugin, could we do it on the main branch first to reduce the delta?

Same for the other minor syntax changes below, e.g.
`from components.java`
vs
`from(components["java"])`
etc. To reduce the risk and make it easier to review, all unrelated improvements should be extracted and be done in the main branch first.

File: build.gradle [hunk 6]
```diff
@@ -8,16 +8,15 @@ plugins {
 	id 'maven-publish'
 	id 'signing'
 	id 'org.ajoberstar.grgit' version '4.1.1'
-	id 'io.codearte.nexus-staging' version '0.30.0'
-	id 'de.marcphilipp.nexus-publish' version '0.4.0'
-	id 'com.github.node-gradle.node' version '7.0.1'
+	id "io.github.gradle-nexus.publish-plugin" version '1.3.0'
```
Comment (Thilo-Acrolinx) [inline build.gradle:11]: please use single quotes here too for consistency

File: .github/workflows/release.yml (hunk 2 above)
Comment (willynganga) [reply .github/workflows/release.yml:46]: Coming to think of it, I don't see any value of using a custom action just to run commands that we can easily do with `./gradlew`. As a matter of fact, it just adds unnecessary maintenance. What do you think?

Comment (Thilo-Acrolinx) [reply .github/workflows/release.yml:46]: Well, we could also use ./gradlew, it should work too. But we should align it, not mix both. Since we use the gradle-build-action already in many GitHub repos and workflows, we should then remove it everywhere.

The gradle-build-action is from the official gradle team on GitHub, it is not just some custom hackish action form a random single dude on the internet. They also document on their page why one should use it:
https://github.com/gradle/gradle-build-action#why-use-the-gradle-build-action

The maintenance is mostly done by Dependabot, so not much work for us. To reduce the number of PRs we get from Dependabot to update this action, we could even switch from the SHA1 hash to the version tag, so then we would only have to update that every leap year or so.

File: build.gradle [hunk 7]
```diff
@@ -225,49 +209,31 @@ publishing {
...
-    username = sonatypeUsername
-    password = sonatypePassword
 }
 
+// Configure Nexus
 nexusPublishing {
     connectTimeout = Duration.ofMinutes(5);
     clientTimeout = Duration.ofMinutes(5);
+
+    transitionCheckOptions {
```
Comment (Thilo-Acrolinx) [inline build.gradle:219]: according to https://github.com/gradle-nexus/publish-plugin#groovy-dsl-1 the default is 60 retries with 10 seconds delay, do we need to override these default values?

File: .github/workflows/release-stable.yml [hunk 8]
```diff
@@ -54,33 +26,55 @@ jobs:
...
-      - name: Validate Gradle wrapper
-        uses: gradle/wrapper-validation-action@56b90f209b02bf6d1deae490e9ef18b21a389cd4
-
       - name: Gradle build
         uses: gradle/gradle-build-action@842c587ad8aa4c68eeba24c396e15af4c2e9f30a
         with:
           arguments: build
 
+      - name: Remove Snapshot Suffix
+        id: removeSnap
```
Comment (Thilo-Acrolinx) [inline .github/workflows/release-stable.yml:35]: > Snap is a software packaging and deployment system developed by Canonical for operating systems that use the Linux kernel and the systemd init system.

Better call this "removeSnapshot" to avoid any misunderstandings.

File: .github/workflows/release-stable.yml [hunk 9]
```diff
@@ -54,33 +26,55 @@ jobs:
...
+          git commit -m "Release v${{ steps.removeSnap.TAGNAME }}"
+
       - name: Create Javadoc
         uses: gradle/gradle-build-action@842c587ad8aa4c68eeba24c396e15af4c2e9f30a
         with:
           arguments: javadoc
 
       - name: Deploy artifacts
-        id: deploy
         if: github.ref == 'refs/heads/main'
```
Comment (Thilo-Acrolinx) [inline .github/workflows/release-stable.yml:52]: what if we want to release from another branch than main?

File: .github/workflows/release-stable.yml [hunk 10]
```diff
@@ -54,33 +26,55 @@ jobs:
...
+      - name: Remove Snapshot Suffix
+        id: removeSnap
+        uses: gradle/gradle-build-action@842c587ad8aa4c68eeba24c396e15af4c2e9f30a
+        with:
+          arguments: removeSnapshotSuffix
+
+      - name: Remove Snapshot Suffix
+        run: |
+          git config --global user.name 'acrolinx-machine-user-for-integrations'
+          git config --global user.email 'danjela.lura@acrolinx.net'
```
Comment (Thilo-Acrolinx) [inline .github/workflows/release-stable.yml:43]: We should not hard-code Danjela's email and use that for every release commit

File: .github/workflows/release-stable.yml [hunk 11]
```diff
@@ -54,33 +26,55 @@ jobs:
...
 
       - name: Deploy artifacts
-        id: deploy
         if: github.ref == 'refs/heads/main'
-        run: ./deploy.sh
+        uses: gradle/gradle-build-action@842c587ad8aa4c68eeba24c396e15af4c2e9f30a
+        with:
+          arguments: publishToSonatype closeAndReleaseStagingRepository --stacktrace
 
       - name: Bump version and push tag
```
Comment (Thilo-Acrolinx) [inline .github/workflows/release-stable.yml:57]: What version is being "bumped" here? I think this action only creates and pushes a Git tag?
Btw, if we already use raw git commands, maybe we could also run a simple `git tag ${{ steps.removeSnap.TAGNAME }}` to create the tag and a `git push origin ${{ steps.removeSnap.TAGNAME }}` to push it? Then we don't need this action here at all?

File: build.gradle [hunk 12]
```diff
@@ -61,6 +60,59 @@ sonar {
     }
 }
 
+def getFixAndFullVersion() {
```
Comment (Thilo-Acrolinx) [inline build.gradle:63]: I would prefer we don't use the "def" keyword and instead specify the actual type (e.g. `String[]`) instead (just like we don't use the `var` keyword in Java).

File: build.gradle [hunk 13]
```diff
@@ -61,6 +60,59 @@ sonar {
...
+        println("Current build version: $currentVersion")
+        def (lastVersionNumber, currentVersionParts) = getFixAndFullVersion()
+        lastVersionNumber = lastVersionNumber.replace('-SNAPSHOT', '')
+
+        def newVersion = currentVersionParts.take(2).join(".") + "." + lastVersionNumber
+        println("New build version: ${newVersion}")
+
+        setVersionProperty(newVersion)
+
+        exec {
```
Comment (Thilo-Acrolinx) [inline build.gradle:88]: since we need to do this multiple times, could we create a small helper function to which we pass the name and value as arguments?

File: build.gradle [hunk 14]
```diff
@@ -61,6 +60,59 @@ sonar {
...
 
+def getFixAndFullVersion() {
+    def currentVersion = currentVersion.split("\\.")
+    def lastVersionNumber = currentVersion[currentVersion.length - 1]
+
+    return [lastVersionNumber, currentVersion]
+}
+
+def setVersionProperty(String newVersion) {
+    ant.propertyfile(file: "gradle.properties") {
```
Comment (Thilo-Acrolinx) [inline build.gradle:71]: Should we really use Apache Ant here? This adds another external dependency (Apache Ant) to the already complicated toolchain and dependency list, cannot we do this with basic Gradle logic instead?

File: .github/workflows/release-stable.yml (hunk 9 above)
Comment (willynganga) [reply .github/workflows/release-stable.yml:52]: This is up for debate, should we release to Maven Central from a branch that is not main?

File: .github/workflows/release-stable.yml (hunk 10 above)
Comment (willynganga) [reply .github/workflows/release-stable.yml:43]: This will be used for testing purposes only, it's not final yet since we need to create a service account that can be used to make the intermediate commits.

File: build.gradle (hunk 12 above)
Comment (willynganga) [reply build.gradle:63]: I believe in Groovy this is how you do it. Let me change it then.

File: build.gradle (hunk 14 above)
Comment (willynganga) [reply build.gradle:71]: Apache ant is already being used by Gradle and is not a new dependency. This is the only suitable solution that updates the version and does not reformat the text each time.

File: build.gradle (hunk 12 above)
Comment (willynganga) [reply build.gradle:63]: Also, note that this method returns multiple values. I use it to access the values directly. To change this I would have to return a string instead which inmho I don't think is very clean.

File: build.gradle (hunk 14 above)
Comment (Thilo-Acrolinx) [reply build.gradle:71]: Are you sure? I don't think we currently use Ant in any of our Gradle builds/scripts? Or am I overlooking something?

File: build.gradle (hunk 12 above)
Comment (Thilo-Acrolinx) [reply build.gradle:63]: `return [ foo, bar ]` in Groovy means you return List, right? Technically your "def" means the method returns Object, which could be anything. I am not a Groovy expert, but maybe a better solution would be to declare a class with the fields we need (and their types well specified), and then return an instance of this class instead of Object/List?

File: build.gradle (hunk 14 above)
Comment (willynganga) [reply build.gradle:71]: Have a look https://docs.gradle.org/current/userguide/ant.html

Comment (Thilo-Acrolinx) [reply build.gradle:71]: I know that using Ant in Gradle builds is possible, however I don't think we do use it in any of our Gradle builds so far? Hence my question if we really should start doing it now for something "simple" like replacing a property value in a property file? Could not that be done without Ant too?

Comment (willynganga) [reply build.gradle:71]: > Apache ant is already being used by Gradle and is not a new dependency. This is the only suitable solution that updates the version and does not reformat the text each time.

Kindly refer to the above reply, the other options that I had looked at were reformatting the whole file which will always result in a diff that is unnecessary.
